adr/keycloak-adr.md

```diff
@@ -0,0 +1,236 @@
+
+This design not only automates the user deprovisioning process but also establishes an event-driven architecture that minimizes tight coupling between services. By leveraging NATS within a Kubernetes environment, we ensure that our architecture is both scalable and resilient, capable of adapting to the evolving needs of our infrastructure and services.
+
+```mermaid
+graph TD
+    A[Security Enhancement with Event-Driven Architecture] --> B[Decoupling Keycloak and GitLab]
+    A --> C[Asynchronous Communication & Queuing]
+    A --> D[Improved Fault Tolerance & Reliability]
+    A --> E[Guaranteed Delivery & Processing]
+
+    B --> B1[Eliminates Direct Dependencies]
+    B --> B2[Isolates System Components]
+
+    C --> C1[No Immediate Need for Receiver Availability]
+    C --> C2[Messages Persisted Until Processed]
+
+    D --> D1[Service Downtime Does Not Result in Data Loss]
+    D --> D2[Independent Recovery Mechanisms]
+
+    E --> E1[Retries Until Successful Processing]
+    E --> E2[Ensures User Deletion Events Are Handled]
+
+    B1 --> F[Reduced Security Risks]
+    C2 --> G[No Loss of Critical Security Events]
+    D2 --> H[Resilience Against System Failures]
+    E2 --> I[Compliance with Security Policies]
+
+    F --> J[Enhanced Overall Security Posture]
+    G --> J
+    H --> J
+    I --> J
+```
+The provided sequence diagram illustrates a scalable and decoupled architecture for handling user deprovisioning, leveraging NATS for centralized publish/subscribe messaging. This pattern ensures that when a user is deleted or deactivated in Keycloak, the event is published to NATS, from where it is distributed to various service listeners, including a GitLab Service Listener and an Audit Listener. The GitLab Service Listener is responsible for revoking the user's access in GitLab, while the Audit Listener updates the Admin UI to display the user deletion event. This architecture can be easily extended to include additional services by adding more listeners that subscribe to the relevant NATS topics for user status changes.
+```mermaid
+sequenceDiagram
+    participant Keycloak as "Keycloak"
+    participant EventListener as "Keycloak Event Listener"
+    participant NATS as "NATS Messaging System"
+    participant GitLabService as "GitLab Service Listener"
+    participant UIService as "Audit Listener"
+    participant AdminUI as "Admin UI"
+    participant OtherService as "Other Service Listener"
+
+    Keycloak->>EventListener: User Deletion/Deactivation
+    EventListener->>NATS: Publish User Status Change
+    NATS->>GitLabService: Notify User Deletion
+    NATS->>UIService: Update for UI Display
+    NATS->>OtherService: Notify User Deletion (Extendable)
+    GitLabService->>GitLab: Revoke User Access
```

Comment:
suggestion: After revoking the user access, should the keycloak listener can also post an update for the Admin UI. For an async/decoupled update like this, I think we'll want to aggregate all the outcomes back for reporting so we can show that all the expected revocations happened.

Reply:
Maybe this is handled via a viewer on the immutable log, Not 100% sure how this whole solution looks to an admin.

Reply:
Thanks @ericwyles. I agree that updates should be captured for reporting and the Admin UI. 

We don't want direct dependencies on KeyCloak to the Admin UI for the reasons I have mentioned, like reliability( the service might be down), tight coupling, etc. 

We are solving the problem with async event-driven architecture.

>Maybe this is handled via a viewer on the immutable log, Not 100% sure how this whole solution looks to an admin.

This is a future enhancement.

Reply:
I mistyped... I meant to say the Gitlab listener reporting success somewhere after it finishes revoking the user in gitlab. I think as long as that goes to the immutable log and the admin can see it (even in a future enhancement) then my concern is satisfied.